src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/ProxyWhitelist.java

```diff
@@ -75,8 +75,51 @@ public ProxyWhitelist(Collection<? extends Whitelist> delegates) {
         reset(delegates);
     }
 
+    /**
+     * This should only be called from {@link ProxyWhitelist#reset(Collection)}. So we can assume that a write 
+     * lock is held.
+     * @param wrapper the wrapper
+     */
+    private void addWrapper(ProxyWhitelist wrapper) {
+        assert wrapper.lock.writeLock().isHeldByCurrentThread();
+        lock.writeLock().lock();
+        try {
+            wrappers.put(wrapper, null);
+            lock.readLock().lock();
+        } finally {
+            lock.writeLock().unlock();
+        }
+        try {
+            for (Whitelist subdelegate : delegates) {
+                if (subdelegate instanceof EnumeratingWhitelist) {
+                    ((EnumeratingWhitelist) subdelegate).clearCache();  // We only care about top-level cache
+                    continue; // this is handled specially
+                }
+                wrapper.delegates.add(subdelegate);
+            }
+            wrapper.methodSignatures.addAll(methodSignatures);
+            wrapper.newSignatures.addAll(newSignatures);
+            wrapper.staticMethodSignatures.addAll(staticMethodSignatures);
+            wrapper.fieldSignatures.addAll(fieldSignatures);
+            wrapper.staticFieldSignatures.addAll(staticFieldSignatures);
+        } finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    /**
+     * This only exists for consistency. Access to originalDelegates should require a read lock and this 
+     * method should be called by a <b>wrapped</b> instance's {@link ProxyWhitelist#reset(Collection)}.
+     */
     private void reset() {
-        reset(originalDelegates);
+        Collection<? extends Whitelist> delegatesCollection;
+        this.lock.readLock().lock();
+        try {
+            delegatesCollection = originalDelegates;
+        } finally {
+            lock.readLock().unlock();
+        }
+        reset(delegatesCollection);
```

Comment:
`reset(Collection)` already acquires a write lock internally, so IMO it would be clearer to acquire the write lock here before calling the method rather than introducing a temporary read lock just to reassign a variable. That way you avoid consistency issues with something else possibly modifying `delegatesCollection`/`originalDelegates` (which are the same object) after the read lock here is dropped but before the write lock in `reset(Collection)` is acquired: 

```suggestion
        lock.writeLock().lock();
        try {
            reset(originalDelegates);
        } finally {
            lock.writeLock().unlock();
        }
```

Reply:
Ah, reentrant lock. Sure, that makes sense.

Reply:
Sound good but we have to check if we are still holding the write lock.
https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.WriteLock.html#unlock()

```suggestion
        lock.writeLock().lock();
        try {
            reset(originalDelegates);
        } finally {
            if (lock.writeLock().isHeldByCurrentThread()) {
                lock.writeLock().unlock();
            }
        }
```

Reply:
I think it's ok without that check, since `lock` increases a counter, and `unlock` decreases the counter and only actually releases the lock if the counter is 0. In this case we'll acquire the lock in `reset()`, the count will go up 1 and then down 1 in `reset(Collection)`, and then the count will go back down to 0 and be released in the finally block in `reset()`. I might be misunderstanding your concern though.

Reply:
Ah, I see. Agreed.

Reply:
It is the middle of the night for Allan, I'm going to commit Devin's suggestion to keep us moving forward.